libafl/src/inputs/bytessub.rs

```diff
@@ -27,10 +30,199 @@ fn end_index<R>(range: &R, max_len: usize) -> usize
+        };
+
+    let end = match (outer_range.end_bound(), inner_range.end_bound()) {
+        (Bound::Unbounded, Bound::Unbounded) => Bound::Unbounded,
+        (Bound::Excluded(bound), Bound::Unbounded) => Bound::Excluded(*bound),
+        (Bound::Unbounded, Bound::Excluded(bound)) => Bound::Excluded(outer_range.end - *bound),
+        (Bound::Included(bound), Bound::Unbounded) => Bound::Included(*bound),
+        (Bound::Unbounded, Bound::Included(bound)) => Bound::Included(outer_range.end - *bound),
+        (Bound::Included(own), Bound::Included(other)) => {
+            Bound::Included(min(*own, outer_range.start + other))
+        }
+        (Bound::Included(own), Bound::Excluded(other)) => {
+            Bound::Included(min(*own, outer_range.start + other - 1))
+        }
+        (Bound::Excluded(own), Bound::Included(other)) => {
+            Bound::Included(min(*own - 1, outer_range.start + other))
+        }
+        (Bound::Excluded(own), Bound::Excluded(other)) => {
+            Bound::Excluded(min(*own, outer_range.start + other))
+        }
+    };
+
+    (start, end)
+}
+
+/// An immutable contiguous subslice of a byte slice.
+/// It is mostly useful to cheaply wrap a subslice of a given input.
+///
+/// A mutable version is available: [`BytesSliceMut`].
+#[derive(Debug)]
+pub struct BytesSlice<'a> {
+    /// The (complete) parent input we will work on
+    parent_slice: OwnedSlice<'a, u8>,
+    /// The range inside the parent input we will work on
+    range: Range<usize>,
+}
+
+/// A mutable contiguous subslice of a byte slice.
+/// It is mostly useful to cheaply wrap a subslice of a given input.
+///
+/// An immutable version is available: [`BytesSlice`].
+#[derive(Debug)]
+pub struct BytesSliceMut<'a> {
+    /// The (complete) parent input we will work on
+    parent_slice: OwnedMutSlice<'a, u8>,
+    /// The range inside the parent input we will work on
+    range: Range<usize>,
+}
+
+impl<'a> BytesSlice<'a> {
+    /// Creates a new [`BytesSlice`], a sub-slice representation of a byte array.
+    pub fn new<R>(parent_slice: OwnedSlice<'a, u8>, range: R) -> Self
+    where
+        R: RangeBounds<usize>,
+    {
+        let parent_len = parent_slice.len();
+
+        BytesSlice {
+            parent_slice,
+            range: Range {
+                start: start_index(&range),
+                end: end_index(&range, parent_len),
+            },
+        }
+    }
+
+    /// Get the sub slice as bytes.
+    #[must_use]
+    pub fn bytes(&self) -> &[u8] {
+        &self.parent_slice[self.range.clone()]
+    }
+
+    /// Creates a new [`BytesSlice`] that's a sliced view on a bytes slice.
+    pub fn with_slice<R>(parent_slice: &'a [u8], range: R) -> Self
+    where
+        R: RangeBounds<usize>,
+    {
+        Self::new(parent_slice.into(), range)
+    }
+
+    /// The parent input
+    #[must_use]
+    pub fn parent_slice(&self) -> OwnedSlice<'a, u8> {
```

Comment:
Why return a clone? Could be `(self) -> OwnedSlice`

Reply:
The user can still chose to `clone()` beforehand

Reply:
my mistake here, i thought `OwnedSlice::clone` was returning the `Ref` variant, not the `Owned` one.
the idea was to return a cheap OwnedSlice and let user decide to perform a 'real' clone or not, without consuming the object.